seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/resourcemanager/AbstractResourceManager.java

```diff
@@ -0,0 +1,273 @@
+                    applyResources(jobId, needApplyResource).whenComplete((s, e) -> {
+                        if (e != null) {
+                            releaseAllResourceInternal(jobId, slotProfiles.values());
+                            completableFuture.completeExceptionally(e);
+                            return;
+                        }
+                        for (int i = 0; i < s.size(); i++) {
+                            slotProfiles.put(needApplyIndex.get(i), s.get(i));
+                            if (successCount.incrementAndGet() == resourceProfile.size()) {
+                                completableFuture.complete(Lists.newArrayList(slotProfiles.values()));
+                            }
+                        }
+
+                    });
+                } else {
+                    releaseAllResourceInternal(jobId, slotProfiles.values());
+                    for (int i = 0; i < slotProfiles.size(); i++) {
+                        if (slotProfiles.get(i) == null) {
+                            completableFuture.completeExceptionally(new NoEnoughResourceException("can't apply resource request: " + resourceProfile.get(i)));
+                            break;
+                        }
+                    }
+                }
+            }
+        });
+        return completableFuture;
+    }
+
+    protected boolean supportDynamicWorker() {
+        return false;
+    }
+
+    private void releaseAllResourceInternal(long jobId, Collection<SlotProfile> slotProfiles) {
+        LOGGER.warning("apply resource not success, release all already applied resource");
+        slotProfiles.stream().filter(Objects::nonNull).forEach(profile -> {
+            releaseResource(jobId, profile);
+        });
+    }
+
+    /**
+     * find new worker in third party resource manager, it returned after worker register successes.
+     *
+     * @param resourceProfiles the worker should have resource profile list
+     */
+    protected void findNewWorker(List<ResourceProfile> resourceProfiles) {
+        throw new UnsupportedOperationException("Unsupported operation to find new worker in " + this.getClass().getName());
+    }
+
+    private <E> InvocationFuture<E> sendToMember(Operation operation, Address address) {
+        InvocationBuilder invocationBuilder =
+                nodeEngine.getOperationService().createInvocationBuilder(SeaTunnelServer.SERVICE_NAME,
+                        operation, address);
+        return invocationBuilder.invoke();
+    }
+
+    @Override
+    public CompletableFuture<Void> releaseResources(long jobId, List<SlotProfile> profiles) {
+        CompletableFuture<Void> completableFuture = new CompletableFuture<>();
+        List<CompletableFuture<Void>> futures = new ArrayList<>();
+        for (SlotProfile profile : profiles) {
+            futures.add(releaseResource(jobId, profile));
+        }
+        CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).whenComplete((r, e) -> {
```

Comment:
I add a test about `CompletableFuture.allOf().whenComplete()`

```

public class TestFuture {
    public static void main(String[] args) {
        TestFuture tf = new TestFuture();
        tf.test();
    }

    public void test() {
        CompletableFuture<Void> uCompletableFuture = CompletableFuture.supplyAsync(() -> {
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
            System.out.println("1111111111111");
            return null;
        });

        CompletableFuture<Void> objectCompletableFuture = CompletableFuture.supplyAsync(() -> {
            System.out.println("22222222222222");
            return null;
        });

        CompletableFuture<Void> objectCompletableFuture1 = CompletableFuture.supplyAsync(() -> {
            System.out.println("333333333333333");
            return null;
        });

        CompletableFuture<Void> objectCompletableFuture2 = CompletableFuture.supplyAsync(() -> {
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
            System.out.println("error1");
            throw new RuntimeException("1");
        });

        CompletableFuture<Void> objectCompletableFuture3 = CompletableFuture.supplyAsync(() -> {
            System.out.println("error2");
            throw new RuntimeException("2");
        });

        List<CompletableFuture<Void>> list = new ArrayList<>();
        list.add(uCompletableFuture);
        list.add(objectCompletableFuture);
        list.add(objectCompletableFuture1);
        list.add(objectCompletableFuture2);
        list.add(objectCompletableFuture3);

        CompletableFuture<Void> voidCompletableFuture = CompletableFuture.allOf(list.toArray(new CompletableFuture[list.size()]));

        voidCompletableFuture.whenComplete((v, t) -> {
            if (t != null) {
                System.out.println(t.getMessage());
            }
        });
    }
}

```
It output message:

```
22222222222222
333333333333333
error2

```

It seems some future not complete.

```
        CompletableFuture<Void> uCompletableFuture = CompletableFuture.supplyAsync(() -> {
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
            System.out.println("1111111111111");
            return null;
        });
```
In your case, If one releaseResource(jobId, profile) throw a Exception, the other profile may can not release.

Reply:
You miss `join()` in the last